### PR TITLE
CD-2725 Generate stats directly to mbtiles

### PR DESCRIPTION
**Pull request by kudlav**
CD-2725

## Objective
Add the ability to generate json directly into mbtiles.

## Description
- Add an option`--into-md` to insert generated tilestats into mbtiles metadata
- `buildGeoStats` rejects promise instead of throwing an error (index.js rewritten into await/async)
- Split geostats and validator tests (structuring the tests)

## Acceptance
- added two tests:
  - option --into-md with non MBTiles file to test error handling
  - MBTiles with option --into-md; which tests writing, writing when stats already exist; reading the written stats

## Checklist

- [X] I have added relevant info to the CHANGELOG.md

### CODE REVIEW

**Review by lazaa32 — CHANGES_REQUESTED** [review]
Can you please update CLI section in README with information about new options?

**Comment by lazaa32** [comment]
Getting an error:
```
dam@XPS:~/dev/maptiler/mapbox-geostats$ ./bin/mapbox-geostats ~/data/mbtiles/cartiqo2.mbtiles
/home/adam/dev/maptiler/mapbox-geostats/index.js:16
  options.maxValuesToReport ??= 50;
                             ^

SyntaxError: Unexpected token '?'
    at Object.compileFunction (vm.js:344:18)
    at wrapSafe (internal/modules/cjs/loader.js:1048:15)
    at Module._compile (internal/modules/cjs/loader.js:1082:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1138:10)
    at Module.load (internal/modules/cjs/loader.js:982:32)
    at Function.Module._load (internal/modules/cjs/loader.js:875:14)
    at Module.require (internal/modules/cjs/loader.js:1022:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/home/adam/dev/maptiler/mapbox-geostats/bin/mapbox-geostats:6:18)
    at Module._compile (internal/modules/cjs/loader.js:1118:30)

```

**Comment by lazaa32** [comment]
> Getting an error:
> 
> ```
> dam@XPS:~/dev/maptiler/mapbox-geostats$ ./bin/mapbox-geostats ~/data/mbtiles/cartiqo2.mbtiles
> /home/adam/dev/maptiler/mapbox-geostats/index.js:16
>   options.maxValuesToReport ??= 50;
>                              ^
> 
> SyntaxError: Unexpected token '?'
>     at Object.compileFunction (vm.js:344:18)
>     at wrapSafe (internal/modules/cjs/loader.js:1048:15)
>     at Module._compile (internal/modules/cjs/loader.js:1082:27)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1138:10)
>     at Module.load (internal/modules/cjs/loader.js:982:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:875:14)
>     at Module.require (internal/modules/cjs/loader.js:1022:19)
>     at require (internal/modules/cjs/helpers.js:72:18)
>     at Object.<anonymous> (/home/adam/dev/maptiler/mapbox-geostats/bin/mapbox-geostats:6:18)
>     at Module._compile (internal/modules/cjs/loader.js:1118:30)
> ```

It seems that upgrading node and npm solved this issue.

**Comment by kudlav** [comment]
You were probably using node v15 or lower. We are now targeting node v16+ which is from 2021.

Readme updated, version of the required node mentioned in the readme

File: lib/tile-analyze.js [hunk 1]
```diff
@@ -44,102 +45,169 @@ class TileAnalyzeStream extends Transform {
...
-module.exports = function(filePath, options) {
-  options = options || {};
-  const stats = createStats();
-  const layerMap = {};
+module.exports = async function tileAnalyze(filePath, options) {
+  const source = await getSource(filePath);
+
+  const info = await getMetadata(source);
+  if (info.tilestats && validator(info.tilestats)) {
+    if (options.intoMd) throw new Error('Tilestats already exists, use without --into-md');
```
Comment (lazaa32) [inline lib/tile-analyze.js]: Could you please make it more descriptive? E.g. "Tilestats already exist. See metadata table, record/row json" or similar. Thanks :)

File: lib/tile-analyze.js [hunk 2]
```diff
@@ -44,102 +45,169 @@ class TileAnalyzeStream extends Transform {
...
+    return info.tilestats;
+  }
+
+  const tilestats = await analyzeSourceStream(source, options);
+
+  if (options.intoMd) {
+    const report = reportStats(tilestats, options);
+    const metadata = { tilestats: report };
+    if (info.vector_layers) metadata.vector_layers = info.vector_layers;
```
Comment (lazaa32) [inline lib/tile-analyze.js:71]: There could be two scenarios, even though just in theory:
- there is missing `json` record in metadata completely
- there is `json` record in metadata but it does not have `vector_layers` in it. 

Could we print two different warnings for these two cases?

Comment (lazaa32) [reply lib/tile-analyze.js:71]: I am also getting this error with MBTiles with no `json` record in metadata table:
```
./bin/mapbox-geostats ~/data/mbtiles/cartiqo3.mbtiles --into-md
Missing vector_layers in metadata json!
TypeError: Cannot read properties of undefined (reading 'size')
    at module.exports (/home/adam/dev/maptiler/mapbox-geostats/lib/report-stats.js:22:37)
    at buildGeoStats (/home/adam/dev/maptiler/mapbox-geostats/index.js:32:10)
```

Comment (lazaa32) [reply lib/tile-analyze.js:71]: Ok, I see the tool expects `size` from `vector_layers` in https://github.com/maptiler/mapbox-geostats/blob/CD-2725_insert-stats/lib/report-stats.js#L22
Maybe missing `json` or `vector_layers` should end up with error not just warning.

**Review by lazaa32 — CHANGES_REQUESTED** [review]

File: lib/tile-analyze.js (hunk 1 above)
Comment (kudlav) [reply lib/tile-analyze.js]: Can be: _Tilestats already exist in json record of metadata table, run without --into-md to display them_

File: lib/tile-analyze.js (hunk 2 above)
Comment (kudlav) [reply lib/tile-analyze.js:71]: I'm using `getInfo` function from mapbox mbtiles library and I am unable to detect if the whole json row is missing or just vector_layers. https://github.com/mapbox/node-mbtiles/blob/master/lib/mbtiles.js#L260-L298
I'd have to query the DB to distinguish this, which I'd prefer not to do.

The second thing `TypeError: Cannot read properties of undefined (reading 'size')` is really weird, can you send me `cartiqo3.mbtiles` to investigate it?

File: lib/tile-analyze.js (hunk 1 above)
Comment (lazaa32) [reply lib/tile-analyze.js]: Sounds good!

File: lib/tile-analyze.js (hunk 2 above)
Comment (lazaa32) [reply lib/tile-analyze.js:71]: Ok, it's a kind of edge-case, do not make difference between missing `json` or `vector_layers`.

Comment (lazaa32) [reply lib/tile-analyze.js:71]: You can test it on any mbtiles and just run `REPLACE INTO metadata ('json', '') WHERE name='json';`
Or you can use these mbtiles: https://drive.google.com/file/d/1WdJIVGx3L-ECY2Rha7-1ywChfyqqvDNL/view?usp=drive_link

Comment (kudlav) [reply lib/tile-analyze.js:71]: Hmm, still cannot reproduce the error with undefined `size` you mention above. 
Tried different scenarios and it behaves as expected (crashing on @mapbox/mbtiles library when parsing).

`json` row missing, writes into created json row with a warning:
```
Missing vector_layers in metadata json!
```
---
`json` is set to `NULL`:
```
/home/vladan/dev/mapbox-geostats/node_modules/@mapbox/mbtiles/lib/mbtiles.js:271
                Object.keys(jsondata).reduce(function(memo, key) {
                       ^
TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at /home/vladan/dev/mapbox-geostats/node_modules/@mapbox/mbtiles/lib/mbtiles.js:271:24
    at Array.forEach (<anonymous>)
    at Statement.<anonymous> (/home/vladan/dev/mapbox-geostats/node_modules/@mapbox/mbtiles/lib/mbtiles.js:262:24)
```
---
`json` is set to empty string:
```
SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at /home/vladan/dev/mapbox-geostats/node_modules/@mapbox/mbtiles/lib/mbtiles.js:269:43
    at Array.forEach (<anonymous>)
    at Statement.<anonymous> (/home/vladan/dev/mapbox-geostats/node_modules/@mapbox/mbtiles/lib/mbtiles.js:262:24)
```
---
`json` is set to `''`:
```
SyntaxError: Unexpected token ' in JSON at position 0
    at JSON.parse (<anonymous>)
    at /home/vladan/dev/mapbox-geostats/node_modules/@mapbox/mbtiles/lib/mbtiles.js:269:43
    at Array.forEach (<anonymous>)
    at Statement.<anonymous> (/home/vladan/dev/mapbox-geostats/node_modules/@mapbox/mbtiles/lib/mbtiles.js:262:24)
```

Comment (lazaa32) [reply lib/tile-analyze.js:71]: Oh, yes, you are right with the mbtiles I've provided I get the same error `SyntaxError: Unexpected end of JSON input`.
The original cartiqo3.mbtiles files is quite big (over 7GB). I will reduce its size and upload it to GDrive.

Comment (kudlav) [reply lib/tile-analyze.js:71]: Nice catch! The error was caused by layerCount check. If the layerCount was zero, it behaved wrong. Fixed in e19bfaff1c239d1141da246c830eb0d6b406cb9e

**Comment by lazaa32** [comment]
@kudlav testing on the same file `cartiqo_small.mbtiles` I get this error:

```
./bin/mapbox-geostats ~/data/mbtiles/pedro.mbtiles cartiqo_small.mbtiles --into-md
Missing vector_layers in metadata json!
Error: Error writing metadata: Error: SQLITE_ERROR: no such column: grid_id
    at tileAnalyze (/home/adam/dev/maptiler/mapbox-geostats/lib/tile-analyze.js:76:13)
    at async buildGeoStats (/home/adam/dev/maptiler/mapbox-geostats/index.js:29:7)

```

**Comment by kudlav** [comment]
You are analyzing `~/data/mbtiles/pedro.mbtiles`, not the `cartiqo_small.mbtiles`

**Review by lazaa32 — APPROVED** [review]
Awesome job. Thank you.
